### PR TITLE
Highlight search match

### PR DESCRIPTION
**Pull request by Nagasaki45**
This is to solve issue #58.

The solution is based on two major changes, described below.

# Query returns `SearchResult`s

First the `query` module, instead of returning the matching entries, yields `models.SearchResult` objects. These got `entry` and `match` field. The entry is the usual dict. The match is a dict with a similar structure but with keys only for values that match the search and values that equal to the substring that was matched. For example, a match can look like this:
```
match = {
    "key": ["surname"],
    "fields": {
        "author": ["Mr", "Surname"],
        "title": ["complex", "model"],
    },
}
```
There are lists as values because multiple substring of an entry can match when using mutiple search terms, and we want to highlight all of them.

# Highlighting the `list` output

`bibo list` uses the `SearchResult`s to highlight the substring in the header / citation that matches the search, and if there are none, add include the fields that were matched in the output.

### CODE REVIEW

**Comment by Nagasaki45** [comment]
This ended up being much more complicated than I expected. @Kappers, would really appreciate your code review here.

File: bibo/query.py [hunk 1]
```diff
@@ -1,81 +1,96 @@
...
 
 
-def _is_matching(entry, search_term):
+def _match(entry, search_term):
+    """
+    Return a similar structure to an entry (nested dict) with matching strings
+    as values.
+    """
+    if search_term.lower() in entry["key"].lower():
+        return {"key": internals.match_case(search_term, entry["key"])}
```
Comment (Kappers) [inline bibo/query.py:36]: Why end here? I mean, this is a good indication of match, but not a very useful/interesting piece of information potentially compared to what you get from e.g. a `key` _and_ `title` match.

File: bibo/internals.py [hunk 2]
```diff
@@ -192,3 +195,50 @@ def complete_path(ctx, args, incomplete):
...
+    Return `text` with sub-string `highlight` in bold.
+    """
+    text = re.sub(
+        highlight, "{}{}{}".format(_ansi_bold, highlight, _ansi_unbold), text,
+    )
+    # Drop ANSI unbold followed by ANSI bold
+    return text.replace(_ansi_unbold + _ansi_bold, "")
+
+
+def highlight_match(text, result, extra_match_info=None):
```
Comment (Kappers) [inline bibo/internals.py:227]: Docstring! I found this a tad confusing at first (before going on to look at `_recursive_search` etc). For example, knowing off the bat that `text` is a string and `result` is `models.SearchResult` which has nested results is useful.

File: bibo/internals.py [hunk 3]
```diff
@@ -192,3 +195,50 @@ def complete_path(ctx, args, incomplete):
...
+        if isinstance(vals, collections.abc.Mapping):
+            inner_result = models.SearchResult(result.entry[key], result.match[key],)
+            text, extra_match_info = highlight_match(
+                text, inner_result, extra_match_info,
+            )
+        else:
+            for val in vals:
+                if val in text:
+                    text = highlight_text(text, val)
+                else:
```
Comment (Kappers) [inline bibo/internals.py:240]: I'm not sure I get why these are exclusive?

File: bibo/models.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import collections
+
+SearchResult = collections.namedtuple("SearchResult", ["entry", "match"])
```
Comment (Kappers) [inline bibo/models.py:3]: Maybe you'll disagree, but `_nested_append` and `_update_result` from `bibo.query` could move here (or somewhere similar, but not `bibo query`) to leave search/matching for that module. If we need a bit more wrapped up in `SearchResult`, maybe its time for `attrs`: https://www.attrs.org/en/stable/why.html#namedtuples

File: bibo/internals.py [hunk 5]
```diff
@@ -192,3 +195,50 @@ def complete_path(ctx, args, incomplete):
...
+    """
+    index = target.lower().find(substring.lower())
+    if index < 0:
+        msg = "Failed to match case: {} not in {}".format(substring, target)
+        raise ValueError(msg)
+    return target[index : (index + len(substring))]
+
+
+_ansi_bold = "\033[1m"
+_ansi_unbold = "\033[22m"
```
Comment (Kappers) [inline bibo/internals.py:213]: Eh, conventions. Maybe something capsy and at the top of the file? Or, is this allowed in python-styling conventions these days?

File: bibo/internals.py [hunk 6]
```diff
@@ -192,3 +195,50 @@ def complete_path(ctx, args, incomplete):
...
+
+def highlight_text(text, highlight):
+    """
+    Return `text` with sub-string `highlight` in bold.
+    """
+    text = re.sub(
+        highlight, "{}{}{}".format(_ansi_bold, highlight, _ansi_unbold), text,
+    )
+    # Drop ANSI unbold followed by ANSI bold
+    return text.replace(_ansi_unbold + _ansi_bold, "")
```
Comment (Kappers) [inline bibo/internals.py:224]: Why is this necessary? highlight == '', or quirk of re-sub? If that's the case, I suppose it should return out earlier.

File: bibo/internals.py [hunk 7]
```diff
@@ -192,3 +195,50 @@ def complete_path(ctx, args, incomplete):
...
     return options
+
+
+def match_case(substring, target):
+    """
+    Return the string `substring` with the same case as in the `target`
+    string.
+    """
+    index = target.lower().find(substring.lower())
+    if index < 0:
```
Comment (Kappers) [inline bibo/internals.py:206]: Eh, I'm not sure how I feel about `match_case` raising. What if there are multiple matches here? For example, this might be used on abstract fields right? If this is going to just find a single hit, which we know already exists (implied based on usage), maybe this should be renamed - `substring_orig_case` or something.

File: tests/bibo/test_internals.py [hunk 8]
```diff
@@ -74,3 +74,43 @@ def test_complete_path(tmpdir):
...
+def test_match_case():
+    assert internals.match_case("hello", "Hello World") == "Hello"
+    assert internals.match_case("WORLD", "Hello World") == "World"
+    with pytest.raises(ValueError):
+        internals.match_case("bibo", "Hello World")
+
+
+def test_highlight_text():
+    s1 = "hello world"
+    s2 = "hello {}w{}orld".format(internals._ansi_bold, internals._ansi_unbold)
```
Comment (Kappers) [inline tests/bibo/test_internals.py:88]: Maybe this should just be a utility, especially given the private-ish internal `internals._ansi_bold`. Like, `format_bold`

File: tests/bibo/test_query.py [hunk 9]
```diff
@@ -49,10 +49,58 @@ def test_search_with_capitalized_search_term(data):
...
+
+
+def test_search_match_details(data):
+    results = list(query.search(data, ["tolkien", "hobbit"]))
+    assert len(results) == 1
+    assert "tolkien" in results[0].match["key"]
+    assert ("title", ["Hobbit"]) in results[0].match["fields"].items()
+
+
+def test_match(data):
```
Comment (Kappers) [inline tests/bibo/test_query.py:68]: For sake of reader, might be nice to copy this raw entry inline to help follow the test - even if its obvious enough, given, ya know, LOTR hehe.

File: bibo/bibo.py [hunk 10]
```diff
@@ -98,21 +97,31 @@ def _list_format_pattern(entries, format_pattern):
         click.echo(internals.format_entry(entry, format_pattern))
 
 
-def _list_citations(entries, database, bibstyle, verbose):
-    entries = [e for e in entries if e["type"] != "string"]
```
[the remark this answers is not in the document]
Comment (Nagasaki45) [reply bibo/bibo.py:101]: Removing this line has nothing to do with the current change. I was modifying this line anyway so decided to delete it. In the past the search wasn't filtering out string typed entries. These are not bibliographic entries but something similar to bibtex variables. Now it does, so this line is redundant.

File: bibo/bibo.py [hunk 11]
```diff
@@ -97,21 +97,31 @@ def _list_format_pattern(entries, format_pattern):
...
+
+        text = "\n".join([header, citation])
+        text, extra_match_info = internals.highlight_match(text, result)
+
+        click.echo(text)
+
+        if extra_match_info:
+            click.secho("Search matched by", underline=True)
+        for key, val in extra_match_info.items():
+            click.echo("{}: {}".format(key, val))
```
[the remark this answers is not in the document]
Comment (Nagasaki45) [reply bibo/bibo.py:124]: We can consider keeping only a few characters from each side of the highlight and truncate the rest. It might be better though to see how it performs at the moment and then decide.

File: bibo/query.py (hunk 1 above)
Comment (Nagasaki45) [reply bibo/query.py:36]: You are absolutely right! This is the first check as searching by key can sometime contradict searching by specific field and value. For example, if the key `krappers:2020` appears in the database, you want the search `bibo list krappers:2020` to handle this as a key first, and not as an entry with the field `krappers` equals to 2020.
Anyway, the rest of the match should be filled and highlighted. Will push a fix later.

File: bibo/internals.py (hunk 2 above)
Comment (Nagasaki45) [reply bibo/internals.py:227]: Agree! Will add. Was also thinking about adding type annotations and use [mypy](http://mypy-lang.org/) check in CI. What do you think?

File: bibo/internals.py (hunk 3 above)
Comment (Nagasaki45) [reply bibo/internals.py:240]: If the part of the match already appears in the header / citation (`text` here) you can just highlight that and there's no need to present extra info explaining the match. The `else` block is there to include highlighted info only when it's not part of the `text`. For example, when an entry matches because of something in the abstract.

File: bibo/models.py (hunk 4 above)
Comment (Nagasaki45) [reply bibo/models.py:3]: I generally prefer not to mix logic and data structures too much. One reason is imports. If `models.py` doesn't depend on any `internals` functionality it is easier to import. Will be happy to restructure some of `query` / `internals` in a more sensible way though. Regarding `attrs`, would check this out. Don't see the benefit at the moment, as our use case is very simple.

File: bibo/internals.py (hunk 5 above)
Comment (Nagasaki45) [reply bibo/internals.py:213]: _AGREE :-)
Will do.

File: bibo/internals.py (hunk 6 above)
Comment (Nagasaki45) [reply bibo/internals.py:224]: The main reasons I think was to write more straightforward tests. You want to compare the string after `highlight_text` with something you make yourself in the test case that looks meaningful. Was also thinking about suggesting some patches to `click` itself with smarter styling. Would you like to test how bibo behaves without this line?

File: bibo/internals.py (hunk 7 above)
Comment (Nagasaki45) [reply bibo/internals.py:206]: You raised some good points here! I'm happy with renaming the function. ATM it matches only the first occurrence. We can improve this to always return a list with all matches, and then there's also no reason to raise. What do you think?

File: tests/bibo/test_internals.py (hunk 8 above)
Comment (Nagasaki45) [reply tests/bibo/test_internals.py:88]: I don't think I understand you. Do you mean changing `internals.highlight_text` to `internals._highlight_text`? Yeah, I'm happy with that. In general we are far from having any reliable API and can make most of `internals` private.

File: tests/bibo/test_query.py (hunk 9 above)
Comment (Nagasaki45) [reply tests/bibo/test_query.py:68]: It's a general problem with the current test suite. It relies quite a lot on the `test.bib` file, and is hard to read without looking at it all the time. We will have to do something about it at one point but prefer to deal with it altogether, I think...

**Comment by Nagasaki45** [comment]
Thanks @Kappers for the comments! Really valuable!!! Going to submit some patches today / tomorrow. Looking forwards to hear what you think after using it a bit.

File: tests/bibo/test_internals.py (hunk 8 above)
Comment (Kappers) [reply tests/bibo/test_internals.py]: I mean, moving this common line which wraps a string with `internals._ansi_bold` into a dedicated utility, if its going to be used all over the place like this.

File: bibo/internals.py (hunk 7 above)
Comment (Kappers) [reply bibo/internals.py:209]: I feel like returning a list (al matches), and just using something like `re.findall`, might be better here. I feel like that's what we want to happen?

Comment (Nagasaki45) [reply bibo/internals.py:209]: Yep! Improving this now.

File: tests/bibo/test_internals.py (hunk 8 above)
Comment (Nagasaki45) [reply tests/bibo/test_internals.py]: Done!

**Comment by Nagasaki45** [comment]
Hi @Kappers! Thanks for all the input!!! I'm going to push it to master now and see how it goes :-) Looking forwards to your feedback.
